feature/syscall-and-fd-operations

Hamza-Bin-Aamir opened the pull request.

SAreebZaheer committed:
feat(syscall): implement custom syscall registration mechanism

Implement custom system call registration using kprobes and syscall table hooking.
This fulfills Branch 2 requirements from the project development plan.

Features implemented:
- Syscall table lookup using kprobes for modern kernels (>= 5.7)
- Write protection management for syscall table modification
- Custom kproxy_enable syscall with CAP_NET_ADMIN capability checking
- Architecture-specific syscall number allocation (x86_64, i386, ARM64)
- Input validation and userspace data copying
- Proper cleanup and syscall unregistration on module unload
- Userspace test program for syscall invocation
- Enhanced test script with syscall testing

Technical details:
- Uses kprobe to find kallsyms_lookup_name on kernels >= 5.7
- Temporarily disables/enables CR0 write protection
- Validates proxy configuration parameters
- Logs syscall invocations with PID/UID tracking

Testing:
- Module compiles successfully for kernel 6.8.0
- Test framework includes syscall enable/disable testing
- Includes userspace C program (test_syscall.c) for validation

SAreebZaheer committed:
docs(syscall): add comprehensive branch 2 implementation summary

Add detailed documentation for syscall registration branch including:
- Implementation details and code references
- Testing procedures and results
- Security considerations
- Known limitations and next steps
- Complete file change summary

SAreebZaheer committed:
docs(readme): update project status for branch 2 completion

Update README.md to reflect:
- Branch 2 (syscall-registration) completion
- New syscall testing instructions
- Updated project structure with new files
- Feature list with v0.2.0 additions
- Current development status

Hamza-Bin-Aamir committed:
docs: update docs to reflect new approach

Hamza-Bin-Aamir committed:
feat(kernel): add syscall

Hamza-Bin-Aamir committed:
feat: Implement basic syscall entry point with capability checks

Hamza-Bin-Aamir committed:
feat: define file descriptor flags and ioctl command structure

Hamza-Bin-Aamir committed:
feat: Add per-fd private data structure

- Define mutex_proxy_context for per-fd state
- Add configuration and statistics storage
- Implement reference counting for context
- Add spinlock for thread-safe access
- Store owner process credentials
- Add connection tracking table

Each fd will have its own independent proxy configuration.

Hamza-Bin-Aamir committed:
feat: Implement anonymous inode creation

- Use anon_inode_getfd() to create fd (like eventfd)
- Handle O_CLOEXEC and O_NONBLOCK flags
- Associate context with fd via private_data
- Set fd name to "[mutex_proxy]" for /proc/self/fd
- Add basic file_operations structure

Anonymous inodes avoid need for character device nodes.

Hamza-Bin-Aamir committed:
feat: Implement file release operation

- Add mutex_proxy_release() for fd close
- Disable proxy on fd close
- Release reference to context
- Add noop_llseek (no seeking on proxy fd)

Ensures proper cleanup when fd is closed.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): implement read() for statistics

- Add mutex_proxy_read() to return current statistics
- Thread-safe read with spinlock
- Returns struct mutex_proxy_stats to userspace
- Supports partial reads

Userspace can read(fd) to get proxy statistics.

Hamza-Bin-Aamir committed:
feat: Implement read() operation for status/stats

Hamza-Bin-Aamir committed:
feat(mutex_proxy): implement write() for configuration

- Add mutex_proxy_write() to accept configuration
- Validate proxy type, port, and version
- Atomic configuration update with spinlock
- Returns bytes written on success

Userspace can write(fd, &config) to configure proxy.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): implement ioctl() operations

- Add ENABLE/DISABLE proxy control
- Add SET_CONFIG/GET_CONFIG for atomic updates
- Add GET_STATS for statistics retrieval
- Support both 32-bit and 64-bit ioctl
- Return -ENOTTY for unknown commands

Provides alternative to read/write for control operations.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): implement poll() for event notifications

Add poll/select/epoll support to the mutex_proxy file descriptor:
- Add wait_queue_head_t and event_count to context structure
- Initialize wait queue in context allocation
- Implement poll() handler that returns:
  * POLLIN|POLLRDNORM (always readable - stats available)
  * POLLOUT|POLLWRNORM (always writable - config accepted)
  * POLLHUP (when proxy disabled)
- Register poll() in file_operations

This completes the standard file operation interface, enabling async I/O
notifications for userspace applications.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): complete syscall implementation with fd creation

Wire up the mutex_proxy_create syscall with full fd creation:
- Syscall implementation moved to linux/kernel/mutex_proxy.c
- Integrated into kernel build system (linux/kernel/Makefile)
- Added kernel header (linux/include/linux/mutex_proxy.h)
- Module updated to reference kernel implementation
- Syscall now functional: allocates context, creates fd, returns to userspace
- File operations complete: read, write, ioctl, poll, release

The syscall is now fully operational when the kernel is built.
Uses syscall number 451 from previous commit.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): handle fd inheritance across fork/exec

Implement proper POSIX fd inheritance semantics:
- Respect O_CLOEXEC flag (close on exec when set)
- Allow fd inheritance in child processes via fork
- Children inherit shared proxy configuration and state
- Add mutex_proxy_applies_to_current() helper function
- Support GLOBAL flag for system-wide proxy application
- Reference counting ensures context lifetime management

Inheritance behavior:
- Without CLOEXEC: fd inherited by children, proxy stays enabled
- With CLOEXEC: fd closed on exec, proxy disabled
- GLOBAL flag: proxy applies to all system processes
- Supports SCM_RIGHTS passing via Unix domain sockets

This completes proper Unix fd semantics for the proxy control fd,
enabling process-level proxy inheritance patterns.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): add ARM64 architecture support

Add ARM64 syscall wiring for mprox_create (syscall 471):
- Update ARM64 64-bit syscall table (syscall_64.tbl)
- Update ARM64 32-bit compatibility table (syscall_32.tbl)
- Follow ARM64 syscall conventions and alignment
- Maintain compatibility with x86_64 syscall numbering

ARM64 support includes:
- Native 64-bit ARM (aarch64) support
- 32-bit ARM application compatibility on 64-bit kernels
- Proper calling conventions for ARM architecture

This enables mutex_proxy on ARM64 platforms including:
- ARM servers (e.g., AWS Graviton, Ampere)
- ARM development boards (e.g., Raspberry Pi 4/5)
- ARM mobile/embedded systems with Linux

Hamza-Bin-Aamir committed:
feat(mutex_proxy): add comprehensive error handling and validation

Implement robust error handling for mutex_proxy syscall:
- Add is_valid_proxy_config() helper for comprehensive validation
- Add NULL pointer checks to all file operations
- Add bounds checking and overflow protection
- Improve error messages with detailed context
- Add resource limit enforcement

Validation coverage:
- NULL pointer validation for all function parameters
- Configuration structure validation (version, type, port, address)
- Buffer size and count validation
- Integer overflow protection
- Position pointer validation
- Resource limits (connection table size)

Error handling improvements:
- Use unlikely() optimization hints for error paths
- Consistent error logging with pr_err/pr_warn
- Clear error codes for all failure modes
- Detailed error messages for debugging

Security hardening:
- Prevent integer overflows in buffer operations
- Validate all user-supplied data before use
- Resource exhaustion protection
- Defense against malformed input

This makes the syscall resistant to fuzzing attacks and
provides clear diagnostics for troubleshooting issues.

Hamza-Bin-Aamir committed:
feat(mutex_proxy): complete Branch 2 - syscall and fd operations

Commits 15, 16, 17:

**Commit 15: Module cleanup and initialization**
- Add module_init() and module_exit() functions
- Add MODULE_LICENSE("GPL"), MODULE_AUTHOR, MODULE_DESCRIPTION
- Reference counting already complete from prior commits
- RCU-safe context destruction already implemented
- Module logs initialization and unload events

**Commit 16: Debugging and logging infrastructure**
- Add module parameter 'debug' for debug logging (default: false)
- Add module parameter 'conn_table_size' (default: 1024, max: 65536)
- Create mpx_debug/info/warn/err logging macros
- Replace all pr_* calls with new macros
- All log messages prefixed with 'mutex_proxy:'
- Debug logging controlled by module parameter

**Commit 17: Documentation and testing**
- Add comprehensive syscall API documentation (docs/SYSCALL_API.md)
- Document all file operations: read(), write(), ioctl(), poll(), close()
- Document fd inheritance semantics (fork, exec, SCM_RIGHTS)
- Document module parameters and usage examples
- Add testing guide (docs/TESTING.md)
- Include unit tests for all operations
- Add manual testing procedures
- Document security considerations and error handling

Branch 2 Complete:
- All 17 commits of Branch 2 (feature/syscall-and-fd-operations) implemented
- Syscall number 471 allocated for x86_64 and ARM64
- Full file descriptor implementation with all operations
- Comprehensive error handling and validation
- Module infrastructure with parameters and logging
- Complete API documentation and testing guide

Next: Branch 3 (feature/userspace-interface) - implement userspace library

Hamza-Bin-Aamir committed:
chore: submodule update

SAreebZaheer commented on docs/BRANCH_PLAN.md:
```diff
@@ -16,70 +16,84 @@ Building a loadable kernel module (LKM) for Linux that provides a system call to
+### 2. `feature/syscall-and-fd-operations`
+**Description:** Implement single system call that returns file descriptor for proxy control
+- Add new system call in the linux kernel.
+- Create `mutex_proxy_create()` syscall that returns a file descriptor
```
Should we change name to kproxy_create()? may cause confusion with actual MUTEX implementation and for another person looking at it

Hamza-Bin-Aamir replied:
oh yes, if you look at [this commit](https://github.com/Hamza-Bin-Aamir/cs311-linux/commit/e6ea99cd6b188598c9553639927266458409910a), I ended up using the name sys_mprox_create and mprox_create to avoid confusion with the mutex terminology.

Will update docs tho

SAreebZaheer commented on docs/SYSCALL_API.md:
```diff
@@ -0,0 +1,384 @@
+
+## Overview
+
+The `mutex_proxy_create()` syscall provides a file descriptor-based interface for kernel-level proxy control. It follows the Unix "everything is a file" philosophy, similar to `eventfd()`, `timerfd()`, and `signalfd()`.
```
Same naming problem here

Hamza-Bin-Aamir committed:
docs: update syscall naming to use mprox_create consistently

- Use mprox_create instead of mutex_proxy_create in documentation
- Add note explaining syscall is named mprox_create for brevity
- Update all code examples to clarify they use mprox_create syscall
- Add note that fd appears as [mutex_proxy] in /proc/<pid>/fd/
- Update related syscalls section to include mprox_create
- Maintain consistency: syscall name is mprox_create, fd name is [mutex_proxy]

Hamza-Bin-Aamir commented:
@SAreebZaheer this commit has the requested naming convention updates

SAreebZaheer commented on src/module/kproxy.c:
all mentions of kproxy need to be converted to mprox

Hamza-Bin-Aamir committed:
refactor(module): remove kproxy.c and update to mutex_proxy naming

- Remove obsolete kproxy.c (old implementation)
- Update Makefile to build mutex_proxy.ko instead of kproxy.ko
- Rename all kproxy references to mutex_proxy throughout
- Update test_syscall.c to test mprox_create syscall (471)
- Update test program to use new fd-based API
- Update test_module.sh to use mutex_proxy naming
- Fix gcc-12 hardcoded compiler issue (use system default)

Module now builds successfully as mutex_proxy.ko with single source file
(mutex_proxy_core.c) containing the loadable module wrapper for the
kernel's built-in mprox_create syscall from Branch 2.
